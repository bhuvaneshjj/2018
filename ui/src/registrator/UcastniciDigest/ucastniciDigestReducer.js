import {
  csStringSortMethod,
  narozeniSortMethod,
  prijmeniJmenoNarozeniSortMethod
} from '../../entities/ucastnici/ucastniciReducer';

export const SortDirTypes = { NONE: 'none', ASC: 'asc', DESC: 'desc' };

const reverseSortDirType = sortDirType => {
  switch (sortDirType) {
    case SortDirTypes.ASC:
      return SortDirTypes.DESC;
    case SortDirTypes.DESC:
      return SortDirTypes.ASC;
    default:
      return SortDirTypes.ASC;
  }
};

export const initialState = {
  isFetching: false,
  kategorieVykonuFilter: '',
  textFilter: '',
  sortColumn: undefined,
  sortDir: SortDirTypes.NONE
};

const ucastniciDigestReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_UCASTNICI_REQUEST':
      return { ...state, isFetching: true };
    case 'FETCH_UCASTNICI_SUCCESS':
    case 'FETCH_UCASTNICI_ERROR':
      return { ...state, isFetching: false };
    case 'UCASTNICI_DIGEST_KATEGORIE_VYKONU_FILTER_CHANGE':
      if (state.kategorieVykonuFilter === action.typKategorie) {
        return { ...state, kategorieVykonuFilter: '' };
      }
      return { ...state, kategorieVykonuFilter: action.typKategorie };
    case 'UCASTNICI_DIGEST_TEXT_FILTER_CHANGE':
      return { ...state, textFilter: action.textFilter.toLowerCase() };
    case 'UCASTNICI_DIGEST_SORT_DIR_CHANGE':
      if (state.sortColumn !== action.sortColumn) {
        return { ...state, sortColumn: action.sortColumn, sortDir: SortDirTypes.ASC };
      }
      return { ...state, sortDir: reverseSortDirType(state.sortDir) };
    default:
      return state;
  }
};

export default ucastniciDigestReducer;

export const getVykony = (kategorie, ucastnik) => {
  const vykony = {};
  const typyKategorie = {};

  ucastnik.roky.forEach(rok => {
    const { vykon } = ucastnik[rok];
    if (vykon) {
      const typKategorie = kategorie[vykon.kategorie].typ;
      vykony[rok] = {
        kategorie: typKategorie,
        dokonceno: vykon.dokonceno
      };
      // Mark this particular typKategorie in typyKategorie dictionary.
      typyKategorie[typKategorie] = typKategorie;
    }
  });

  return { vykony, typyKategorie };
};

export const getUcastniciDigestSorted = ({
  kategorie,
  ucastnici,
  kategorieVykonuFilter,
  textFilter,
  sortColumn,
  sortDir
}) => {
  const result = [];
  ucastnici.allIds.forEach(id => {
    const ucastnik = ucastnici.byIds[id];
    const posledniUcast = ucastnik[ucastnik.roky[0]];
    const { udaje } = posledniUcast;

    if (
      udaje.prijmeni.toLowerCase().startsWith(textFilter) ||
      udaje.jmeno.toLowerCase().startsWith(textFilter)
    ) {
      const { vykony, typyKategorie } = getVykony(kategorie, ucastnik);
      typyKategorie[''] = true; // empty filter means a match as well
      if (typyKategorie[kategorieVykonuFilter]) {
        result.push({
          id,
          prijmeni: udaje.prijmeni,
          jmeno: udaje.jmeno,
          narozeni: udaje.narozeni,
          ...vykony
        });
      }
    }
  });

  const sortMethods = {
    prijmeni: (a, b) => csStringSortMethod(a.prijmeni, b.prijmeni),
    jmeno: (a, b) => csStringSortMethod(a.jmeno, b.jmeno),
    narozeni: (a, b) => narozeniSortMethod(a.narozeni, b.narozeni)
  };

  const sortMethod = sortMethods[sortColumn] || prijmeniJmenoNarozeniSortMethod;
  const sorted = result.sort(sortMethod);
  if (sortDir === SortDirTypes.DESC) {
    sorted.reverse();
  }

  return sorted.map(ucastnik => {
    const { narozeni, ...ostatek } = ucastnik;
    const { den, mesic, rok } = narozeni;
    return { ...ostatek, narozeni: mesic && den ? `${den}. ${mesic}. ${rok}` : `${rok}` };
  });
};
