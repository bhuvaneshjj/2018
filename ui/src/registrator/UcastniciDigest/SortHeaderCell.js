import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Cell } from 'fixed-data-table-2';
import { SortDirTypes } from './ucastniciDigestReducer';

const sortDirs = {};
sortDirs[SortDirTypes.NONE] = '';
sortDirs[SortDirTypes.ASC] = '↑';
sortDirs[SortDirTypes.DESC] = '↓';

class SortHeaderCell extends Component {
  sortDirChanged = event => {
    event.preventDefault();

    this.props.onSortDirChange();
  };

  render = () => (
    <Cell>
      <a onClick={this.sortDirChanged}>
        {this.props.children} {sortDirs[this.props.sortDir]}
      </a>
    </Cell>
  );
}

SortHeaderCell.propTypes = {
  onSortDirChange: PropTypes.func.isRequired,
  sortDir: PropTypes.string,
  children: PropTypes.node
};

export default SortHeaderCell;
