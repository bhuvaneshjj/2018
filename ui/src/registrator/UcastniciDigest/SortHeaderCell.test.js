import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { SortDirTypes } from './ucastniciDigestReducer';
import SortHeaderCell from './SortHeaderCell';

it('renders the default', () => {
  const component = renderer.create(
    <SortHeaderCell onSortDirChange={jest.fn()} sortDir={SortDirTypes.NONE}>
      příjmení
    </SortHeaderCell>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders jméno sestupně', () => {
  const component = renderer.create(
    <SortHeaderCell onSortDirChange={jest.fn()} sortDir={SortDirTypes.DESC}>
      jméno
    </SortHeaderCell>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('maps onClick to dispatch onSortDirChange action', () => {
  const onSortDirChange = jest.fn();
  const wrapper = mount(
    <SortHeaderCell onSortDirChange={onSortDirChange} sortDir={SortDirTypes.ASC}>
      jméno
    </SortHeaderCell>
  );
  expect(wrapper.find('a')).toHaveLength(1);
  wrapper.find('a').simulate('click');

  expect(onSortDirChange).toHaveBeenCalled();
});
