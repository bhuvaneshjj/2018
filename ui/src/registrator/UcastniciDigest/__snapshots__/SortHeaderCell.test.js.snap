// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders jméno sestupně 1`] = `
<div
  className="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1"
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <div
    className="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
  >
    <div
      className="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
    >
      <div
        className="public_fixedDataTableCell_cellContent"
      >
        <a
          onClick={[Function]}
        >
          jméno
           
          ↓
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the default 1`] = `
<div
  className="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1"
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <div
    className="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
  >
    <div
      className="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
    >
      <div
        className="public_fixedDataTableCell_cellContent"
      >
        <a
          onClick={[Function]}
        >
          příjmení
           
          
        </a>
      </div>
    </div>
  </div>
</div>
`;
